src/transformers/modeling_tf_bart.py

```diff
@@ -780,6 +793,57 @@ def call(self, input_ids: tf.Tensor, use_cache=False):
         return super().call(positions + self.offset)  # super object is not callable for some reason
 
 
+class TFSinusoidalPositionalEmbedding(TFSharedEmbeddings):
+    """This module produces sinusoidal positional embeddings of any length."""
+
+    def __init__(self, num_positions, embedding_dim, **kwargs):
+
+        if embedding_dim % 2 != 0:
+            raise NotImplementedError(f"odd embedding_dim {embedding_dim} not supported")
+        super().__init__(
+            num_positions,
+            embedding_dim,
+            **kwargs,
+        )
+        # self.weight = self._init_weight(*self.weight.shape)
+
+    def build(self, input_shape):
+        """
+        Build shared token embedding layer Shared weights logic adapted from
+        https://github.com/tensorflow/models/blob/a009f4fb9d2fc4949e32192a944688925ef78659/official/transformer/v2/embedding_layer.py#L24
+        """
+        super().build(input_shape)  # Instantiates self.weight so it can be loaded
+        weight: np.ndarray = self._init_weight(self.vocab_size, self.hidden_size)
+        self.set_weights([weight])  # overwrite self.weight to correct value
+
+    @staticmethod
+    def _init_weight(n_pos, dim):
+        """
+        Identical to the XLM create_sinusoidal_embeddings except features are not interleaved. The cos features are in
+        the 2nd half of the vector. [dim // 2:]
+        """
+        position_enc = np.array(
+            [[pos / np.power(10000, 2 * (j // 2) / dim) for j in range(dim)] for pos in range(n_pos)]
+        )
+        # index 0 is all zero
+        position_enc[:, 0 : dim // 2] = np.sin(position_enc[:, 0::2])
+        position_enc[:, dim // 2 :] = np.cos(position_enc[:, 1::2])
+        # convert to tensor
+        table = tf.convert_to_tensor(position_enc, dtype=tf.float32)
+        tf.stop_gradient(table)
+        return table
+
+    def call(self, input_ids, use_cache=False):
+        """Input is expected to be of size [bsz x seqlen]."""
+        bsz, seq_len = input_ids.shape[:2]
```

Comment:
better to use `shape_list()`

Reply:
@stas00 the magician do you know how to replace every occurrence of 
`X.shape` with `shape_list(X)` for arbitrary X?

Reply:
sure, but I will need some samples to know how to restrict the search. 

If for example X doesn't contain spaces, then it's easy:
```
perl -pi -e 's|(\S+)\.shape|shape_list($1)|' files
```

Reply:
Hugely helpful, thanks. It went 23/25. Only issues were multiple in same line.
x is a python variable. @patrickvonplaten  this may be useful to you in future!

Reply:
oh, just add `|g` for global:
```
perl -pi -e 's|(\S+)\.shape|shape_list($1)|g' files
```
glad to hear it worked.